spark-version/1.5-plus/src/main/scala/ZippedPartitionsWithLocalityRDD.scala

```diff
@@ -62,7 +62,7 @@ class ZippedPartitionsWithLocalityRDD[A: ClassTag, B: ClassTag, V: ClassTag](
       while (p < candidateLocs.length) {
         locs = candidateLocs(p)._2.intersect(curPrefs)
         if (!locs.isEmpty) {
-          matchPartition = candidateLocs.remove(p)
+          matchPartition = candidateLocs.apply(p)
```

Comment:
You cannot do this; the logic should be:
1) If partition0 (from rdd(0)) has the same location as partition1 (from rdd(1)), we'll zip them together and remove partition1 from the candidate list
2) If no partition in rdd(1) has the same location as the partition0 from rdd(0), we should randomly select remove one partition from the candidate list and zip it with partition0